### ISSUE TITLE
Cannot publish the biz.aQute.bnd version 4.3.0 plugins because the server says the version already exists

### ISSUE THREAD

**Issue by bjhargrave** [gradle/plugin-portal-requests#9]
### Expected Behavior

Since version 4.3.0 of the plugins are not visible in the plugin repo, I expect the plugin to be published.
### Current Behavior

```
> Task :publishPlugins FAILED
Publishing plugin biz.aQute.bnd.builder version 4.3.0

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':publishPlugins'.
> Failed to post to server.
  Server responded with:
  Artifact [group: biz.aQute.bnd, artifactId: biz.aQute.bnd.gradle,version: 4.3.0, classifier: None, extension: pom] has already been published

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':publishPlugins'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$3.accept(ExecuteActionsTaskExecuter.java:166)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$3.accept(ExecuteActionsTaskExecuter.java:163)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:191)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:156)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:108)
        at org.gradle.api.internal.tasks.execution.ResolveBeforeExecutionOutputsTaskExecuter.execute(ResolveBeforeExecutionOutputsTaskExecuter.java:67)
        at org.gradle.api.internal.tasks.execution.ResolveAfterPreviousExecutionStateTaskExecuter.execute(ResolveAfterPreviousExecutionStateTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:94)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:95)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:43)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:355)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:343)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:336)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:322)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:134)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:129)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:202)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:193)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:129)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: java.lang.RuntimeException: Failed to post to server.
Server responded with:
Artifact [group: biz.aQute.bnd, artifactId: biz.aQute.bnd.gradle,version: 4.3.0, classifier: None, extension: pom] has already been published
        at com.gradle.publish.ResponseUtil.assertValidResponse(ResponseUtil.java:17)
        at com.gradle.publish.OAuthHttpClient.send(OAuthHttpClient.java:78)
        at com.gradle.publish.PublishTask.doSignedPost(PublishTask.java:333)
        at com.gradle.publish.PublishTask.publishToPortal(PublishTask.java:215)
        at com.gradle.publish.PublishTask.publish(PublishTask.java:72)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:103)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:49)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:42)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:28)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:717)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:684)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$5.run(ExecuteActionsTaskExecuter.java:476)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:461)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:444)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.access$200(ExecuteActionsTaskExecuter.java:93)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.execute(ExecuteActionsTaskExecuter.java:237)
        at org.gradle.internal.execution.steps.ExecuteStep.lambda$execute$1(ExecuteStep.java:33)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:33)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:26)
        at org.gradle.internal.execution.steps.CleanupOutputsStep.execute(CleanupOutputsStep.java:58)
        at org.gradle.internal.execution.steps.CleanupOutputsStep.execute(CleanupOutputsStep.java:35)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:48)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:33)
        at org.gradle.internal.execution.steps.CancelExecutionStep.execute(CancelExecutionStep.java:39)
        at org.gradle.internal.execution.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:73)
        at org.gradle.internal.execution.steps.TimeoutStep.execute(TimeoutStep.java:54)
        at org.gradle.internal.execution.steps.CatchExceptionStep.execute(CatchExceptionStep.java:35)
        at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:51)
        at org.gradle.internal.execution.steps.SnapshotOutputsStep.execute(SnapshotOutputsStep.java:45)
        at org.gradle.internal.execution.steps.SnapshotOutputsStep.execute(SnapshotOutputsStep.java:31)
        at org.gradle.internal.execution.steps.CacheStep.executeWithoutCache(CacheStep.java:208)
        at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:70)
        at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:45)
        at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:49)
        at org.gradle.internal.execution.steps.StoreSnapshotsStep.execute(StoreSnapshotsStep.java:43)
        at org.gradle.internal.execution.steps.StoreSnapshotsStep.execute(StoreSnapshotsStep.java:32)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:38)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:24)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:96)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.lambda$execute$0(SkipUpToDateStep.java:89)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:54)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:38)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:76)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:37)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:36)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:26)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:90)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:48)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:69)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:47)
        at org.gradle.internal.execution.impl.DefaultWorkExecutor.execute(DefaultWorkExecutor.java:33)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:140)
        ... 34 more


* Get more help at https://help.gradle.org

BUILD FAILED in 5s
```
### Context

Well I cannot publish newer versions of the Bnd Gradle plugins on the Gradle Plugin portal. So I direct users to use Maven Central or JCenter instead.

### Steps to Reproduce 

I use the script at <https://github.com/bndtools/bnd/blob/master/biz.aQute.bnd.gradle/publish.gradle> to publish the Bnd Gradle plugins to the Gradle Plugin portal so I can publish the same binaries built in our CI build that are on Maven Central and JCenter to the Gradle Plugin portal.
### Your Environment

```
------------------------------------------------------------
Gradle 5.6.3
------------------------------------------------------------

Build time:   2019-10-18 00:28:36 UTC
Revision:     bd168bbf5d152c479186a897f2cea494b7875d13

Kotlin:       1.3.41
Groovy:       2.5.4
Ant:          Apache Ant(TM) version 1.9.14 compiled on March 12 2019
JVM:          1.8.0_222 (Azul Systems, Inc. 25.222-b10)
OS:           Mac OS X 10.15 x86_64
```

**Comment by lacasseio** [comment]
@bjhargrave Could you give it another go? We checked everything on our side and it should succeed now.

**Comment by bjhargrave** [comment]
Same problem exists: 

```
> Task :publishPlugins FAILED
Publishing plugin biz.aQute.bnd.builder version 4.3.1

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':publishPlugins'.
> Failed to post to server.
  Server responded with:
  Artifact [group: biz.aQute.bnd, artifactId: biz.aQute.bnd.gradle,version: 4.3.1, classifier: None, extension: pom] has already been published

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 5s
```

I am now using Gradle 6.0.1.

Interestingly <https://plugins.gradle.org/m2/biz/aQute/bnd/biz.aQute.bnd.gradle/> does not show a 4.3.0 or 4.3.1 directory, but <https://plugins.gradle.org/m2/biz/aQute/bnd/biz.aQute.bnd.gradle/maven-metadata.xml> for some reason includes:

```xml
<version>4.3.0</version>
<version>4.3.1</version>
```

in the `<versions>` section!?

**Comment by lacasseio** [comment]
Thanks for trying it again. It looks like something is wrong.

**Comment by lacasseio** [comment]
We think we found the issue. Here is an link to the internal discussion (Gradle inc only) for the required context: https://gradle.slack.com/archives/CCNH1LKLY/p1574187804005400. You should be able to publish now. Let's reopen again if there is still a problem.

**Comment by bjhargrave** [comment]
Well it seems a partial success...

```
$ ../gradlew --no-daemon -b publish.gradle displayArtifacts publishPlugins
To honour the JVM settings for this build a new JVM will be forked. Please consider using the daemon: https://docs.gradle.org/6.0.1/userguide/gradle_daemon.html.
Daemon will be stopped at the end of the build stopping after processing

> Task :displayArtifacts
DefaultPublishArtifact_Decorated biz.aQute.bnd.gradle-4.3.0:pom:pom:null /Users/hargrave/.gradle/caches/modules-2/files-2.1/biz.aQute.bnd/biz.aQute.bnd.gradle/4.3.0/bc0b98d4941a8015a571a7d9217ddfcd4d7e750e/biz.aQute.bnd.gradle-4.3.0.pom
DefaultPublishArtifact_Decorated biz.aQute.bnd.gradle-4.3.0:jar:jar:null /Users/hargrave/.gradle/caches/modules-2/files-2.1/biz.aQute.bnd/biz.aQute.bnd.gradle/4.3.0/50c6ccb20cebceca2f18aa6fad1ff5359cb75a32/biz.aQute.bnd.gradle-4.3.0.jar
DefaultPublishArtifact_Decorated biz.aQute.bnd.gradle-4.3.0-javadoc:jar:jar:javadoc /Users/hargrave/.gradle/caches/modules-2/files-2.1/biz.aQute.bnd/biz.aQute.bnd.gradle/4.3.0/b0788b729843981309cef55c6ca0eddb6ed756ea/biz.aQute.bnd.gradle-4.3.0-javadoc.jar
DefaultPublishArtifact_Decorated biz.aQute.bnd.gradle-4.3.0-sources:jar:jar:sources /Users/hargrave/.gradle/caches/modules-2/files-2.1/biz.aQute.bnd/biz.aQute.bnd.gradle/4.3.0/6e44c64a21fd72aa9696f1a43d48ed5f2ae746a6/biz.aQute.bnd.gradle-4.3.0-sources.jar

> Task :publishPlugins
Publishing plugin biz.aQute.bnd.builder version 4.3.0
Publishing plugin biz.aQute.bnd version 4.3.0
Publishing plugin biz.aQute.bnd.workspace version 4.3.0
Publishing artifact file:/Users/hargrave/.gradle/caches/modules-2/files-2.1/biz.aQute.bnd/biz.aQute.bnd.gradle/4.3.0/bc0b98d4941a8015a571a7d9217ddfcd4d7e750e/biz.aQute.bnd.gradle-4.3.0.pom
Publishing artifact file:/Users/hargrave/.gradle/caches/modules-2/files-2.1/biz.aQute.bnd/biz.aQute.bnd.gradle/4.3.0/50c6ccb20cebceca2f18aa6fad1ff5359cb75a32/biz.aQute.bnd.gradle-4.3.0.jar
Publishing artifact file:/Users/hargrave/.gradle/caches/modules-2/files-2.1/biz.aQute.bnd/biz.aQute.bnd.gradle/4.3.0/b0788b729843981309cef55c6ca0eddb6ed756ea/biz.aQute.bnd.gradle-4.3.0-javadoc.jar
Publishing artifact file:/Users/hargrave/.gradle/caches/modules-2/files-2.1/biz.aQute.bnd/biz.aQute.bnd.gradle/4.3.0/6e44c64a21fd72aa9696f1a43d48ed5f2ae746a6/biz.aQute.bnd.gradle-4.3.0-sources.jar
Publishing artifact build/publish-generated-resources/pom.xml
Activating plugin biz.aQute.bnd.builder version 4.3.0
Activating plugin biz.aQute.bnd version 4.3.0
Activating plugin biz.aQute.bnd.workspace version 4.3.0

BUILD SUCCESSFUL in 12s

$ ../gradlew --no-daemon -b publish.gradle displayArtifacts publishPlugins
To honour the JVM settings for this build a new JVM will be forked. Please consider using the daemon: https://docs.gradle.org/6.0.1/userguide/gradle_daemon.html.
Daemon will be stopped at the end of the build stopping after processing

> Task :displayArtifacts
DefaultPublishArtifact_Decorated biz.aQute.bnd.gradle-4.3.1:pom:pom:null /Users/hargrave/.gradle/caches/modules-2/files-2.1/biz.aQute.bnd/biz.aQute.bnd.gradle/4.3.1/553fc026c4cbee71f2812ea38ff642a59ddfc4cb/biz.aQute.bnd.gradle-4.3.1.pom
DefaultPublishArtifact_Decorated biz.aQute.bnd.gradle-4.3.1:jar:jar:null /Users/hargrave/.gradle/caches/modules-2/files-2.1/biz.aQute.bnd/biz.aQute.bnd.gradle/4.3.1/7bb3da2fa5cda338862de275acf66c3a4368466e/biz.aQute.bnd.gradle-4.3.1.jar
DefaultPublishArtifact_Decorated biz.aQute.bnd.gradle-4.3.1-javadoc:jar:jar:javadoc /Users/hargrave/.gradle/caches/modules-2/files-2.1/biz.aQute.bnd/biz.aQute.bnd.gradle/4.3.1/66039267b4c662a257f5617ff2cf9d0e500a0dfc/biz.aQute.bnd.gradle-4.3.1-javadoc.jar
DefaultPublishArtifact_Decorated biz.aQute.bnd.gradle-4.3.1-sources:jar:jar:sources /Users/hargrave/.gradle/caches/modules-2/files-2.1/biz.aQute.bnd/biz.aQute.bnd.gradle/4.3.1/ab712dd0bd27fa67ae82b3e9d70710a298236094/biz.aQute.bnd.gradle-4.3.1-sources.jar

> Task :publishPlugins
Publishing plugin biz.aQute.bnd.builder version 4.3.1
Publishing plugin biz.aQute.bnd version 4.3.1
Publishing plugin biz.aQute.bnd.workspace version 4.3.1
Publishing artifact file:/Users/hargrave/.gradle/caches/modules-2/files-2.1/biz.aQute.bnd/biz.aQute.bnd.gradle/4.3.1/553fc026c4cbee71f2812ea38ff642a59ddfc4cb/biz.aQute.bnd.gradle-4.3.1.pom
Publishing artifact file:/Users/hargrave/.gradle/caches/modules-2/files-2.1/biz.aQute.bnd/biz.aQute.bnd.gradle/4.3.1/7bb3da2fa5cda338862de275acf66c3a4368466e/biz.aQute.bnd.gradle-4.3.1.jar
Publishing artifact file:/Users/hargrave/.gradle/caches/modules-2/files-2.1/biz.aQute.bnd/biz.aQute.bnd.gradle/4.3.1/66039267b4c662a257f5617ff2cf9d0e500a0dfc/biz.aQute.bnd.gradle-4.3.1-javadoc.jar
Publishing artifact file:/Users/hargrave/.gradle/caches/modules-2/files-2.1/biz.aQute.bnd/biz.aQute.bnd.gradle/4.3.1/ab712dd0bd27fa67ae82b3e9d70710a298236094/biz.aQute.bnd.gradle-4.3.1-sources.jar
Publishing artifact build/publish-generated-resources/pom.xml
Activating plugin biz.aQute.bnd.builder version 4.3.1
Activating plugin biz.aQute.bnd version 4.3.1
Activating plugin biz.aQute.bnd.workspace version 4.3.1

BUILD SUCCESSFUL in 11s
```

However:

1. Only a pom exists on the server. No other artifacts (such as the jar) are present which differs from 4.2.0. https://plugins.gradle.org/m2/biz/aQute/bnd/biz.aQute.bnd.gradle/4.3.1/

2. The pom is not the pom I asked to publish. It is some fabricated pom (`build/publish-generated-resources/pom.xml`) which differs from the actual pom I asked to publish which is the pom in the jar artifact. The fabricated pom does not contain the necessary dependencies of the plugin so people cannot use it directly. See https://github.com/bndtools/bnd/issues/3174. I don't see why the publish task fabricates a pom when I supply the actual pom.

At this point, I am wondering if I should delete these versions (4.3.0 and 4.3.1) from the plugin repository.

**Comment by lacasseio** [comment]
Oh, this is very much a mess. I will have a look today again and see what happened this time. Thank you for you patience.

**Comment by tomdcc** [comment]
@bjhargrave The plugin publishing plugin will always generate its own pom file and upload that. You're seeing an error because you're forcing it to also upload another pom, so it then complains that there already is one.

This is just how the publishing plugin works currently. If you need to customise the dependencies of the generated pom, you can do so with the `withDependencies` block, see [here](https://discuss.gradle.org/t/how-can-i-publish-an-existing-plugin-jar-to-the-plugin-portal/11186/11?u=tomd) for an example.

Since the plugin had uploaded all the artifacts, I've marked them as valid in the db and you can consume 4.3.0 and 4.3.1 now.

It looks like these plugins need dependencies added to the generated pom. To do so you can either add them in the dependencies section of the gradle project above, or manually add them to the pom with the `withDependencies` block. Since it looks like that file isn't doing anything other than publishing the plugin, it's probably ok to add them as dependencies in there.

When you have done this then either publish a 4.3.2 or contact us to remove 4.3.0/1 and try again.

**Comment by bjhargrave** [comment]
I already have a complete and fully valid pom since the gradle plugin artifact is already in maven central. I don't want to duplicate the dependency information in the gradle build script to attempt to regenerate a pom with the right dependency information as this is prone to errors when the gradle plugin artifact's dependencies change from release to release. If there is some way to use the actual pom's real dependencies to populate the plugin publish process' generated pom's dependencies, I could use that. But the `withDepenencies` block is not documented so I have no idea what I could do with it.

My central issue is the need to separate the act of building the gradle plugin artifact from the act of publishing the gradle plugin artifact to the plugin portal. The sticking point is the need to generate a new pom, with proper dependency information, during publication when I already have the complete pom in hand from the earlier building of the gradle plugin artifact.

**Comment by marcphilipp** [comment]
@bjhargrave Please see https://github.com/marcphilipp/gradle-sandbox/blob/master/plugin-maven-publish/build.gradle an example on how to publish the same POM to Maven Central and the Plugin Portal.

**Comment by lacasseio** [comment]
@bjhargrave Could you help us understand the publishing history of your plugin? It seems the change in how the plugin was published happened between version [3.3.0](https://plugins.gradle.org/m2/biz/aQute/bnd/biz.aQute.bnd.gradle/3.3.0/) and [3.4.0](https://plugins.gradle.org/m2/biz/aQute/bnd/biz.aQute.bnd.gradle/3.4.0/). It would mean that starting with version 3.4.0, the POM included with the plugin has being the one you are seeing for version 4.3.0 and 4.3.1.

Did consuming the plugin starting version 3.4.0 worked in any way with Gradle? According to my quick lookup, the consumption of the plugin would have [worked up until version 4.1.0](https://bintray.com/beta/#/bnd/bnd/biz.aQute.bnd.gradle/4.1.0?tab=files), but the POM used would have being the one affected by whatever change was done for version 3.4.0. Starting from [version 4.2.0](https://bintray.com/beta/#/bnd/bnd/biz.aQute.bnd.gradle/4.2.0?tab=files), extra dependencies were added which weren't translated properly to the POM published on the plugin portal.

My best guess is starting from 3.4.0, the publication seems to have switched to the plugin portal native publishing, however, the POM wasn't properly configured. My suspicion are supported by the lack of `asc` and hash files present with version 3.3.0 on the plugin portal. We should have only started to see issues with version 4.2.0 but the publishing failure with 4.3.0 make everyone look more closely to the published files which showed the POM dependency mismatched. Does that make any sense?

What we could do, is delete the plugin 4.3.0 and 4.3.1 again and republish with a fixed dependencies OR revert to however the plugin was published in version 3.3.0 if that is even possible. I realized that plugin 3.3.0 was published in 2016 where things may have changed considerably since then.

**Comment by lacasseio** [comment]
We kept looking into this and it ends up that Bintray synchronization was decommission a while back with rare exception for legacy reason. [Your concern here](https://github.com/bndtools/bnd/issues/2468#issuecomment-395910140) were on the right track. The `mavenCoordinate` block simply overwrite the maven coordinate when publishing to the plugin portal when people use the [legacy plugin application](https://docs.gradle.org/current/userguide/plugins.html#sec:old_plugin_application). The POM and Jar are still published to the plugin portal as seen in the Gradle build. The publication of the plugin will have to be configured as Tom or Marc explained. In fact, it was only coincidental that plugin version 3.4.0 to 4.1.0 worked at they didn't have any dependencies. Plugin version 4.2.0 required a dependency that wasn't configured in the POM due to how the configuration was done.

The[ instruction here](https://github.com/bndtools/bnd/blob/master/biz.aQute.bnd.gradle/publish.gradle#L1-L3) are wrong and you should instead apply the script to your build script to the artifact is picked up properly by the `com.gradle.plugin-publish` plugin. Inside your [build.gradle](https://github.com/bndtools/bnd/blob/master/biz.aQute.bnd.gradle/build.gradle) you should be able to add `apply from: 'publish.gradle'`. If it is important, we can look at removing 4.2.0, 4.3.0 and 4.3.1 given they would actually never work properly. As for the previous version, although it is not the same POM, they are essentially the same given both (Bintray vs Plugin Portal) doesn't contain any dependencies.

**Comment by bjhargrave** [comment]
Let me try to explain. Bnd is a build tool, primarily for OSGi bundles, and Bnd is self hosted in that we use Bnd to build Bnd. As a tool, Bnd includes support for using maven style repositories both consuming artifacts from maven style repositories and publishing to maven style repositories. As part of that support Bnd will generate pom files using the the Bnd build process. So the Bnd build, while driven by Gradle, uses the Bnd Gradle plugin for Bnd workspace models builds to build Bnd itself and publish the build to a maven style repository. Our CI build runs on Azure Pipelines and published to JFrog. You can see the final Bnd 4.3.1 build for the Bnd Gradle plugin was published to https://bndtools.jfrog.io/bndtools/libs-release-local/biz/aQute/bnd/biz.aQute.bnd.gradle/4.3.1/ from the Azure Pipelines build.

So we do not in any way use Gradle plugins for publishing to maven style repositories and generating poms. We are eating our own food to make Bnd and test that Bnd properly generates valid pom files and properly publishes to maven style repositories. From the JFrog published build, I have shell scripts which take the final build artifacts and use REST calls to publish them to JCenter and then Maven Central.

Prior to Bnd 4.2, the Bnd Gradle plugin was  FAT jar in that it included all classes it needed to function. Starting in Bnd 4.2, the Bnd Gradle plugin was modified to use dependencies in its pom file to supply needed classes from other dependencies (such as other Bnd projects). So prior to Bnd 4.2, it did not matter that the gradle portal publish plugin generated a basically empty pom. But for Bnd 4.2 and later, we need to have the proper dependencies expressed in the pom for proper operation. This is why I want to be able to use the already built pom instead remaking the pom during the gradle portal publish plugin process.

**Comment by lacasseio** [comment]
I understand and it make sense what you are trying to achieve. Then I think the right step forward would be "replace" the generated POM with yours. To do so, you could use a solution similar to what [Marc shown](https://github.com/marcphilipp/gradle-sandbox/blob/master/plugin-maven-publish/build.gradle) and use [`MavenPom#withXml` hook](https://docs.gradle.org/current/dsl/org.gradle.api.publish.maven.MavenPom.html#org.gradle.api.publish.maven.MavenPom:withXml(org.gradle.api.Action)) to replace the current XML nodes with the one from your `pom.xml` file. You can use the [Groovy XmlSlurper](https://docs.groovy-lang.org/latest/html/api/groovy/util/XmlSlurper.html).

I will let you open an feature request for the publish plugin to expose an API for using your own `pom.xml` file on this repository. Once you have the right configuration for your plugin `pom.xml`, open another issue with what you want to do with the already published plugins version 4.2 and later.
